File: examples/ruby/sse.rb
#!/usr/bin/env ruby
# SSE consumer for /v1/stream. Requires Trader / Developer / Enterprise tier.
#
#   bundle install
#   export ORUK_API_KEY=ork_xxxxxxxx
#   ruby sse.rb
require 'json'
require 'ld-eventsource'  # gem: ld-eventsource

abort 'Set ORUK_API_KEY first.' unless (key = ENV['ORUK_API_KEY'])

client = SSE::Client.new(
  'https://api.oruk.ai/v1/stream',
  headers: { 'X-API-Key' => key, 'Accept' => 'text/event-stream' },
)

client.on_event do |event|
  data = JSON.parse(event.data) rescue {}
  case event.type
  when 'story'
    puts "[#{(data['id'] || '')[4, 8]}] #{(data['urgency'] || '?').ljust(9)} " \
         "#{(data['category'] || '?').ljust(9)} #{data['headline']}"
  when 'corroboration'
    puts "  ↑ +#{(data['newSource'] || '').ljust(20)} now #{data['count'] || 0} sources on " \
         "#{(data['storyId'] || '')[4, 8]}"
  when 'heartbeat'
    puts "· heartbeat - #{data['activeSources'] || 0} sources live"
  else
    puts "[#{event.type}] #{data.to_json}"
  end
end

client.on_error do |err|
  warn "[sse] error: #{err.inspect} (will auto-reconnect)"
end

# Block forever. ld-eventsource handles reconnect + Last-Event-ID automatically.
loop { sleep 1 }

File: examples/javascript/sse.js
/**
 * SSE consumer for /v1/stream - requires Trader / Developer / Enterprise tier.
 *
 *   npm install
 *   export ORUK_API_KEY=ork_xxxxxxxx
 *   node sse.js
 *
 * Auto-reconnects via the `eventsource` client. The Last-Event-ID header
 * (sent automatically by eventsource on reconnect) resumes from the last
 * successfully-delivered event.
 */
import EventSource from 'eventsource';

const key = process.env.ORUK_API_KEY;
if (!key) {
  console.error('Set ORUK_API_KEY first.');
  process.exit(1);
}

const url = 'https://api.oruk.ai/v1/stream';
const headers = {
  'X-API-Key': key,
  'Accept': 'text/event-stream',
};

// `eventsource@2` does not let us pass arbitrary headers in the browser-like
// constructor, but exposes an `https.options.headers` escape hatch.
const es = new EventSource(url, { headers });

es.onopen = () => process.stderr.write('[sse] connected\n');

es.onerror = err => {
  if (err?.status === 401) { console.error('[sse] 401 unauthorized - check ORUK_API_KEY'); process.exit(1); }
  if (err?.status === 403) { console.error('[sse] 403 forbidden - SSE requires Trader/Developer/Enterprise tier'); process.exit(1); }
  process.stderr.write(`[sse] error - ${err?.message ?? err} (will auto-reconnect)\n`);
};

es.addEventListener('story', ev => {
  const d = JSON.parse(ev.data);
  console.log(
    `[${(d.id ?? '').slice(4, 12)}] ${(d.urgency ?? '?').padEnd(9)} ` +
    `${(d.category ?? '?').padEnd(9)} ${d.headline ?? ''}`
  );
});

es.addEventListener('corroboration', ev => {
  const d = JSON.parse(ev.data);
  console.log(
    `  ↑ +${(d.newSource ?? '').padEnd(20)} now ${d.count ?? 0} sources on ${(d.storyId ?? '').slice(4, 12)}`
  );
});

es.addEventListener('heartbeat', ev => {
  const d = JSON.parse(ev.data);
  console.log(`· heartbeat - ${d.activeSources ?? 0} sources live`);
});

// Graceful shutdown
process.on('SIGINT', () => { es.close(); process.exit(0); });
